Client/Frontend/Home/ActivityStreamTopSitesCell.swift

```diff
@@ -0,0 +1,392 @@
+        }
+    }
+
+}
+
+struct ASHorizontalScrollCellUX {
+    static let TopSiteCellIdentifier = "TopSiteCell"
+    static let TopSiteItemSize = CGSize(width: 100, height: 100)
+    static let BackgroundColor = UIColor.whiteColor()
+    static let PageControlRadius: CGFloat = 3
+    static let PageControlSize = CGSize(width: 30, height: 15)
+    static let PageControlOffset = -5
+}
+
+/*
+ The View that describes the topSite cell that appears in the tableView.
+ */
+class ASHorizontalScrollCell: UITableViewCell {
+
+    lazy private var collectionView: UICollectionView = {
+        let layout  = HorizontalFlowLayout()
+        layout.itemSize = ASHorizontalScrollCellUX.TopSiteItemSize
+        layout.heightShouldChange = { [weak self] num in
+            self?.heightChanged(num)
+        }
+        let collectionView = UICollectionView(frame: CGRectZero, collectionViewLayout: layout)
+        collectionView.registerClass(TopSiteCell.self, forCellWithReuseIdentifier: ASHorizontalScrollCellUX.TopSiteCellIdentifier)
+        collectionView.backgroundColor = ASHorizontalScrollCellUX.BackgroundColor
+        collectionView.showsHorizontalScrollIndicator = false
+        collectionView.pagingEnabled = true
+        return collectionView
+    }()
+
+    lazy private var pageControl: FilledPageControl = {
+        let pageControl = FilledPageControl()
+        pageControl.tintColor = UIColor.grayColor()
+        pageControl.indicatorRadius = ASHorizontalScrollCellUX.PageControlRadius
+        pageControl.userInteractionEnabled = false
+        return pageControl
+    }()
+
+    weak var parentTableView: UITableView?
+
+    override init(style: UITableViewCellStyle, reuseIdentifier: String?) {
+        super.init(style: style, reuseIdentifier: reuseIdentifier)
+
+        backgroundColor = ASHorizontalScrollCellUX.BackgroundColor
+        contentView.addSubview(collectionView)
+        contentView.addSubview(pageControl)
+
+        collectionView.snp_makeConstraints { make in
+            make.edges.equalTo(contentView).priorityLow()
+        }
+
+        pageControl.snp_makeConstraints { make in
+            make.size.equalTo(ASHorizontalScrollCellUX.PageControlSize)
+            make.top.equalTo(collectionView.snp_bottom).offset(ASHorizontalScrollCellUX.PageControlOffset)
+            make.centerX.equalTo(self.snp_centerX)
+        }
+    }
+
+    override func layoutSubviews() {
+        super.layoutSubviews()
+        let layout = collectionView.collectionViewLayout as! HorizontalFlowLayout
+        pageControl.pageCount = layout.numberOfPages
+        pageControl.hidden = pageControl.pageCount <= 1
+    }
+
+    func heightChanged(newHeight: Int) {
+        collectionView.snp_updateConstraints { make in
+            make.edges.equalTo(contentView).priorityLow()
+            make.height.equalTo(newHeight).priorityMedium()
+        }
+        self.layoutSubviews()
+        parentTableView?.reloadData()
```

Comment:
Ah I see now about your comment about the hack in `ActivityStreamPanel` where you assign the parentTableView to the cell. If I understand correctly (let me know if I'm totally off base),  but this is similar in how we handle layout for cells on the old top sites where the cells scale which in turn inform the number of rows/columns that we show. This in turn dictates how high the `ASHoriziontalScrollView` should be thus the need to inform our tableView. I know at least in the old top sites logic this was hard to reason about and deal with.

What if we inversed the relationship? Instead of having this chain of cell -> ASHorizontalScrollCell -> UITableView, what if the table view tells the row, fit in X amount space. The scroll view and cells would then adjust accordingly. I think by having the table view dictate the terms of sizing, it will be easier to update the layout based on view controller lifecycle events and trait collections since the table view and view controller are next together in the hierarchy. Also would give us better control over exactly how we want to render top sites for various phones/layouts/sizes.